onnxruntime/core/providers/cuda/cu_inc/common.cuh

```diff
@@ -360,6 +366,16 @@ __device__ __inline__ double _Min(double a, double b) {
   return (isnan(a) || isnan(b)) ? std::numeric_limits<double>::quiet_NaN() : ( a < b ? a : b );
 }
 
+template <>
+__device__ __inline__ half _Min(half a, half b) {
+  return ISNAN_HALF(a) ? a : (ISNAN_HALF(b) ? b : (a < b ? a : b));
```

Comment:
May use __hmin_nan(a, b) for _Min and __hmax_nan(a, b) for _Max, when a/b is half or bfloat16 (casted to __nv_bfloat16). See https://docs.nvidia.com/cuda/archive/12.1.0/cuda-math-api/group__CUDA__MATH____HALF__COMPARISON.html#group__CUDA__MATH____HALF__COMPARISON_1g9752fee573b47368538ab19495ab9623

Reply:
Thanks, I've added a change to use these functions now. I also updated the tests to test with bfloat16.

Reply:
I'm not sure whether I need to provide a fallback implementation for CUDA < 11.0, does ONNX Runtime still support older CUDA versions?

Reply:
Looks like CUDA 10 support was removed in #12541, so I've removed the check for CUDA_VERSION.

Reply:
Looks like there is build error: 
   error : identifier "__hmax_nan" is undefined 

May need the following:
```
#if __CUDA_ARCH__ >= 530 || !defined(__CUDA_ARCH__)
  return __hmin_nan(a, b);
#else
  return _Min(float(a), float(b));
#endif
```

Reply:
:+1: thanks for all your help with this, I've used your suggestion and also had to define constants for NaN values as `CUDART_NAN_FP16` and `CUDART_NAN_BF16` are only defined since CUDA 12.2.